runner/src/mill/runner/MillMain.scala

```diff
@@ -348,6 +362,69 @@ object MillMain {
     }
   }
 
+  def validatePorts(config: MillCliConfig, streams: SystemStreams): Option[Set[Int]] = {
+    val result =
+      if (config.ports.isDefined) {
+        var ports = Set.empty[Int]
+        try {
+          ports = config.ports.get.split(",").map(_.toInt).toSet
+        } catch {
+          case e: NumberFormatException =>
+            streams.err.println(
+              "Failed to parse  --ports, please provide comma sepperated values. E.g 4090,5000,2910"
+            )
+            None
+        }
+        // Port validation
+        ports.foreach { case port =>
+          // 1024 is the min port on windows
+          if (port < 1024 || port > 65535) {
+            streams.err.println(
+              s"Failed to parse $port. Ports provided must be within 1024 to 65535."
+            )
+            None
+          }
+          try {
+            new ServerSocket(port).close()
+          } catch {
+            case e: BindException => {
+              streams.err.println(
+                s"Failed to bind to $port! Please only provide free ports."
+              )
+              None
+            }
+          }
+        }
+        Some(ports)
+      } else {
+        var i = 0
+        var ports = Set.empty[Int]
+        for (_ <- 1 to 100) {
+          if (i == 6) {
+            return Some(ports)
+          }
+          val socket = new ServerSocket(0)
+          try {
+            val port = socket.getLocalPort
+            if (!ports.contains(port)) {
+              ports = ports + port
+              i += 1
+            }
+
+          } finally {
+            socket.close()
+          }
+        }
+
+        streams.err.println(
+          "Failed to generate enough ports for servers to use. Either there isn't 6 free ports on your computer, or there is a firewall issue. "
```

Comment:
Magic number `6`.

Reply:
> Magic number `6`.

Fixed but, maybe we don't need to do this at all.